### PR TITLE
Added PHPStan, apply changes for level 3

### PR DESCRIPTION
**Pull request by Majkl578**
Mostly annotation changes/fixes to match what the methods/properties really are used for. Added few asserts and instanceof.
Not really sure if you want to consider this as BC issue - I'd hope not, but if so we can retarget to develop instead.

Level 3, not going further as it requires actually changing code itself, that should follow later in develop with strict typing.

There is a PHP notice currently since Travis doesn't have OCI driver:
```
Notice: Use of undefined constant OCI_DEFAULT - assumed 'OCI_DEFAULT'
```

### CODE REVIEW

File: lib/Doctrine/DBAL/Connection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: Wouldn't it be enough to check if the driver is not null? Type hint should help static analysis w/o the run time check.

File: lib/Doctrine/DBAL/Connection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: I'd say it's a bug and should be fixed, not documented.

File: lib/Doctrine/DBAL/Connection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: Boolean is a bug.

File: lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php]: Isn't it something like `(DriverConnection|null)[]`? Not all values are necessary of the same type.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Should it be fixed or silenced instead? Doesn't look like it's by design (the SQLite platform impelementation).

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Looks like a bug. Mixing something with void doesn't make sense.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Can elements of the array be anything else than string? `string[]|string` would make more sense.

File: lib/Doctrine/DBAL/Schema/Schema.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Schema.php]: A nullable type would make more sense here.

File: lib/Doctrine/DBAL/Sharding/PoolingShardConnection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Sharding/PoolingShardConnection.php]: The variable could be initialized with an empty array instead.

**Review by morozov — COMMENTED** [review]
I'd say, most if not all places where `false`/`bool` is added as part of the union type are not by design and should be fixed in a BC way (if possible) or silenced instead of being documented.

**Comment by carusogabriel** [comment]
Already prepared changes for `tests/` as well https://github.com/carusogabriel/dbal/tree/phpstan-tests

File: lib/Doctrine/DBAL/Connection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Connection.php]: `$this->_platform` is typed as AbstractPlatform, but `createDatabasePlatformForVersion()` is only defined in some drivers implementing VersionAwarePlatformDriver.
I at least replaced this by `assert()` inside the `if` since `Connection::getDatabasePlatformVersion()` returns null for non-version-aware drivers.

File: lib/Doctrine/DBAL/Connection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Connection.php]: Hmm this is because `SqlAnywhereConnection::getServerVersion()` calls `ResultStatement::fetchColumn()` which returns `string|bool`. :| Replaced with assert there and removed here.

File: lib/Doctrine/DBAL/Connection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Connection.php]: Same as above.

File: lib/Doctrine/DBAL/Connection.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Connection.php]: This on the other hand is not a bug, some implementors explicitly return `false` (`SQLSrvConnection::errorCode()`).

File: lib/Doctrine/DBAL/Connection.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Connection.php]: Same here (`OCI8Connection::lastInsertId()`).

File: lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php]: Such notation is more correct, but not supported i.e. by IDEs (PhpStorm will break completely). Pretty much same like intersections vs. unions we discussed earlier somewhere else.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: I can change those ints to strings, but I really have no idea of consequences (either in DBAL or for consumers). SqlAnywherePlatoform returns ints too... Maybe better leave this to 3.x?

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: As much as I hate using `false` as type, there is no better way around here. Usually these cases are something to replace with `null` for 3.x.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Yep, removed. But `OracleSchemaManager::createDatabase()` returns true so I removed that one. Should be documented in UPGRADING?

File: lib/Doctrine/DBAL/Sharding/PoolingShardConnection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Sharding/PoolingShardConnection.php]: Probably ok here and for `$connections`, changed (`close()` now sets empty array instead of null).

File: lib/Doctrine/DBAL/Schema/Schema.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/Schema.php]: Can't be done in 2.x I guess...

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: No idea since `$table` is uselessly types as `array`. If you are certain, I can change it.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: I'd look at it from the caller's standpoint: whatever callers can consume by design (`string[]` or `string`) should be valid return, the rest shouldn't.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: I'm all for leaving it for 3.x where we could fix the actual code instead of complicating the API documentation.

**Comment by carusogabriel** [comment]
Any updates on this one? :)

**Comment by Majkl578** [comment]
I'll wait for PHPStan 0.10 (to be released next month) before finishing this.

File: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php
Comment (carusogabriel) [inline lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php]: Does this function always return a `stdClass`? Can't we force it in PHPStan stubs?

File: lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php
Comment (carusogabriel) [inline lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php]: PHPStan stubs false-positive?

Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php]: Probably.

File: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php]: I didn't actually check whether it is stdClass or not, it returns _some object crate_: http://php.net/db2_server_info

Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php]: oh, and `false`, how unusual ;)

File: lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php
Comment (morozov) [reply lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php]: It's actually `OCI-Lob`.

File: lib/Doctrine/DBAL/Connection.php
Comment (morozov) [reply lib/Doctrine/DBAL/Connection.php]: DBAL should eliminate this discrepancy. The clients will not have to know the low level details.

File: lib/Doctrine/DBAL/Driver/Connection.php
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/Connection.php]: `bool|null` should be eliminated, not document.

File: lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php]: Can we change the `_fetch()`'es phpDoc instead?

File: lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php]: `false|null` is undesirable.

File: lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php]: `false|null` is undesirable.

File: lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php]: `false|null` is undesirable.

File: lib/Doctrine/DBAL/Schema/Schema.php
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Schema.php]: Let's ignore it and fix instead of introducing an intermediate incorrect type.

File: lib/Doctrine/DBAL/Schema/Table.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/Table.php]: `string|false`?

File: phpstan.neon.dist
Comment (morozov) [inline phpstan.neon.dist]: Can this be replaced with `assert($statement instanceof PDOStatement)` in the code?

File: phpstan.neon.dist
Comment (morozov) [inline phpstan.neon.dist]: Could we generate a stub in the longer term? As long as we officially support the driver.

File: phpstan.neon.dist
Comment (morozov) [inline phpstan.neon.dist]: Could we generate a stub in the longer term?

File: tests/phpstan-polyfill.php
Comment (morozov) [inline tests/phpstan-polyfill.php]: Is it really `'db2'`, not `'ibm_db2'`?
```php
var_dump(extension_loaded('db2'));
// bool(false)

var_dump(extension_loaded('ibm_db2'));
// bool(true)
```

File: tests/phpstan-polyfill.php
Comment (morozov) [inline tests/phpstan-polyfill.php]: Given that the stub files are nicely organized, in order to avoid copy/paste and simplify maintenance, can this be reworked as:
```php
$extensions = ['ibm_db2', ...];

foreach ($extensions as $extension) {
    if (extension_loaded($extension)) {
        continue;
    }

    require sprintf(__DIR__ . '/../vendor/jetbrains/phpstorm-stubs/%1$s/%1$s.php', $extension);
}
```

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Please ignore instead.

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: Please figure out the correct type from the API standpoint and ignore the deviations.

File: lib/Doctrine/DBAL/Schema/Schema.php
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/Schema.php]: Please ignore.

File: phpstan.neon.dist
Comment (Majkl578) [reply phpstan.neon.dist]: I have no idea. PhpStorm stubs don't have sasql and I didn't find any documentation either, nor PECL extension. Working with SAP SQL Anywhere is really hard. :(

File: lib/Doctrine/DBAL/Schema/Table.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/Table.php]: will ignore

File: lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php]: adding to ignore, but it's a bug, false is explicitly returned here :/

File: lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]:
```
 ------ ------------------------------------------------------------------------------------------------------------------------ 
  Line   lib/Doctrine/DBAL/Schema/OracleSchemaManager.php                                                                        
 ------ ------------------------------------------------------------------------------------------------------------------------ 
  104    Method Doctrine\DBAL\Schema\OracleSchemaManager::_getPortableTableDefinition() should return array but returns string.  
 ------ ------------------------------------------------------------------------------------------------------------------------ 

 ------ ---------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php                                                                        
 ------ ---------------------------------------------------------------------------------------------------------------------------- 
  222    Method Doctrine\DBAL\Schema\PostgreSqlSchemaManager::_getPortableTableDefinition() should return array but returns string.  
 ------ ---------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Doctrine/DBAL/Schema/SQLServerSchemaManager.php                                                                        
 ------ --------------------------------------------------------------------------------------------------------------------------- 
  210    Method Doctrine\DBAL\Schema\SQLServerSchemaManager::_getPortableTableDefinition() should return array but returns string.
```
Honestly, no idea. Some managers return `$table` as is and some return a string (schema name + table name).
This definitely looks like a bad design (git-blame indicates addition back in 2008-2011).
Help? :)

Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php]: putting it into ignores for now, unlikely to be properly fixable in 2.x

File: lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php]: hmm, that's not a valid class name 🙈 moving to ignore rather than typehinting _hack_

File: lib/Doctrine/DBAL/Driver/Connection.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/Connection.php]: ignored

File: phpstan.neon.dist
Comment (Majkl578) [reply phpstan.neon.dist]: Not sure if possible for a class constant...

Comment (morozov) [reply phpstan.neon.dist]: It's missing from the PDO stub as well as many other driver-specific symbols. I'll file a PR for [JetBrains/phpstorm-stubs](https://github.com/JetBrains/phpstorm-stubs) later.

File: phpstan.neon.dist
Comment (morozov) [reply phpstan.neon.dist]: For future reference, I found a few working links at https://archive.sap.com/documents/docs/DOC-40537. I'll try to drop in the Windows DLL into PHP 5 and see if I can dump the signatures. Sometime.

File: phpstan.neon.dist
Comment (morozov) [inline phpstan.neon.dist]: Hmm… all generated stubs have it as `OCI_Lob` (probably to make it valid), shouldn't then PHPStan just use the "fixed" names? The `oci_new_descriptor()` is stubbed as `oci_new_descriptor() : OCI_Lob`, and `OCI_Lob::writeTemporary()` is also stubbed, so I'd expect not seeing this error.

File: phpstan.neon.dist
Comment (morozov) [reply phpstan.neon.dist]: Filed https://github.com/JetBrains/phpstorm-stubs/pull/381.

Comment (morozov) [reply phpstan.neon.dist]: > Not sure if possible for a class constant…

Oh, I see what you're saying. The class stub will be not loaded for PDO in the extension itself is loaded, even if it's compiled w/o the missing constants. PDO is so nicely designed.

Comment (Majkl578) [reply phpstan.neon.dist]: It's definitely missing in the stubs, but it is also missing in PHP itself unless oci extension is loaded. Who said multiple inheritance in PHP is not possible? :grin:

File: phpstan.neon.dist
Comment (Majkl578) [reply phpstan.neon.dist]: https://github.com/phpstan/phpstan/issues/1123

Comment (Majkl578) [reply phpstan.neon.dist]: Not really sure what to think about this. :)

File: phpstan.neon.dist
Comment (Majkl578) [reply phpstan.neon.dist]: The link lead me to: http://dcx.sap.com/index.html#sa160/en/dbprogramming/php-api.html
This looks good enough to build the stubs. Maybe i.e. @carusogabriel may want to help here? :)

But we should probably consider dropping the platform/driver altogether.
Given that DBAL 3 will require PHP 7.2, and SQL Anywhere is barely keeping up (the link only lists PHP 5 modules)...

**Comment by Majkl578** [comment]
CI passed. :green_heart:

File: phpstan.neon.dist
Comment (Majkl578) [reply phpstan.neon.dist]: Nevermind, did this during Slack outage. :stuck_out_tongue: 
https://github.com/JetBrains/phpstorm-stubs/pull/382

**Review by morozov — APPROVED** [review]

**Comment by Majkl578** [comment]
@morozov Do you want this in 2.8 or wait for release and then merge into 2.9?

**Comment by morozov** [comment]
@Majkl578 I understand it as a non-breaking low-risk, mostly documentation improvement. Unless it's not true, I'm fine with having it in `2.8.0`.

**Comment by Majkl578** [comment]
Yup, and no tests changed so it should be fine. Let's ship it. 🚢

**Comment by Ocramius** [comment]
:+1:

**Pull request merged**
